backends/tfhe-cuda-backend/cuda/src/pbs/programmable_bootstrap_multibit.cu

```diff
@@ -379,24 +379,48 @@ void cleanup_cuda_multi_bit_programmable_bootstrap(cuda_stream_t *stream,
 }
 
 // Returns a chunk size that is not optimal but close to
-__host__ uint32_t get_lwe_chunk_size(uint32_t ct_count) {
+template <typename Torus, class params>
+__host__ uint32_t get_lwe_chunk_size(int gpu_index, uint32_t max_num_pbs,
+                                     uint32_t polynomial_size,
+                                     uint32_t max_shared_memory) {
 
-#if CUDA_ARCH >= 900
-  // Tesla H100
-  return (ct_count > 10000) ? 30 : 64;
-#elif CUDA_ARCH >= 890
-  // Tesla RTX4090
-  return 8;
-#elif CUDA_ARCH >= 800
-  // Tesla A100
-  return (ct_count > 10000) ? 30 : 45;
-#elif CUDA_ARCH >= 700
-  // Tesla V100
-  return (ct_count > 10000) ? 12 : 18;
-#else
-  // Generic case
-  return (ct_count > 10000) ? 2 : 1;
-#endif
+  uint64_t full_sm_keybundle =
+      get_buffer_size_full_sm_multibit_programmable_bootstrap_keybundle<Torus>(
+          polynomial_size);
+
+  int max_blocks_per_sm;
+  if (max_shared_memory < full_sm_keybundle)
+    cudaOccupancyMaxActiveBlocksPerMultiprocessor(
+        &max_blocks_per_sm,
+        device_multi_bit_programmable_bootstrap_keybundle<Torus, params, NOSM>,
+        polynomial_size / params::opt, full_sm_keybundle);
+  else
+    cudaOccupancyMaxActiveBlocksPerMultiprocessor(
+        &max_blocks_per_sm,
+        device_multi_bit_programmable_bootstrap_keybundle<Torus, params,
+                                                          FULLSM>,
+        polynomial_size / params::opt, 0);
+
+  int num_sms = 0;
+  check_cuda_error(cudaDeviceGetAttribute(
+      &num_sms, cudaDevAttrMultiProcessorCount, gpu_index));
+
+  int x = num_sms * max_blocks_per_sm;
+  int count = 0;
+
+  int ith_divisor = 0;
+  int log2_max_num_pbs = log2(max_num_pbs);
+  if (log2_max_num_pbs > 14)
+    ith_divisor = log2_max_num_pbs - 14;
```

Comment:
Why 14 here? This part of the code will need a comment explaining why it's done this way I think, otherwise we'll be lost when  coming back to it later. :slightly_smiling_face:

Reply:
I'm empirically looking for some reasonable value. 14 didn't work with scalar_mul. I will try 13.

I added a comment explaining that.